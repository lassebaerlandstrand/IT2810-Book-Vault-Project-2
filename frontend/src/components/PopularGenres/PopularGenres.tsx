import { ReactElement } from 'react';
import {
  IconCompass,
  IconCrystalBall,
  IconGlobe,
  IconHourglassEmpty,
  IconMasksTheater,
  IconPencil,
} from '@tabler/icons-react';
import { Link } from 'react-router-dom';
import { Box, Flex, Paper, ScrollArea, Text, Title } from '@mantine/core';
import styles from './PopularGenres.module.css';

type BookCategoryProps = {
  name: string;
  icon: ReactElement;
  link: string;
};

type BookCategoriesProps = { categories: BookCategoryProps[]; typeOfCategory: string };

const BookCategory = ({ name, icon, link }: BookCategoryProps) => {
  return (
    <Link to={link} className={styles.link} aria-label={`Go to ${name}`}>
      <Paper className={styles.categoryCard}>
        {icon}
        <Text>{name}</Text>
      </Paper>
    </Link>
  );
};

const BookCategories = ({ categories, typeOfCategory }: BookCategoriesProps) => {
  return (
    <Box my="xl">
      <Title order={2} fw={600} mb="sm" size="h3">
        Popular {typeOfCategory}
      </Title>
      <ScrollArea w="100%" type="auto" offsetScrollbars="x" scrollbarSize={6}>
        <Flex justify="space-between" gap="md" py="xs">
          {categories.map((category) => (
            <BookCategory key={category.name} {...category} />
          ))}
        </Flex>
      </ScrollArea>
    </Box>
  );
};

const iconThickness = 1;

const popularCategories = [
  {
    name: 'Drama',
    icon: <IconMasksTheater stroke={iconThickness} aria-label="Drama" />,
    link: '/books?genres=Drama',
  },
  {
    name: 'Historical',
    icon: <IconHourglassEmpty stroke={iconThickness} aria-label="Historical" />,
    link: '/books?genres=Historical',
  },
  {
    name: 'Geography',
    icon: <IconGlobe stroke={iconThickness} aria-label="Geography" />,
    link: '/books?genres=Geography',
  },
  {
    name: 'Fantasy/Fiction',
    icon: <IconCrystalBall stroke={iconThickness} aria-label="Fantasy/Fiction" />,
    link: '/books?genres=Fantasy/Fiction',
  },
  {
    name: 'Adventure',
    icon: <IconCompass stroke={iconThickness} aria-label="Adventure" />,
    link: '/books?genres=Adventure',
  },
  {
    name: 'Literature',
    icon: <IconPencil stroke={iconThickness} aria-label="Literature" />,
    link: '/books?genres=Literature',
  },
];

/**
 * Displays a horizontally scrollable list of popular book genres with icons and links.
 */
const PopularGenres = () => {
  return <BookCategories categories={popularCategories} typeOfCategory="Genres" />;
};

export default PopularGenres;
