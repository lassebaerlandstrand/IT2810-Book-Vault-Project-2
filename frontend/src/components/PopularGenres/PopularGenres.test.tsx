import { render, screen } from '@test-utils';
import { MemoryRouter } from 'react-router-dom';
import PopularGenres from './PopularGenres';

describe('PopularGenres component', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <PopularGenres />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct category names', () => {
    render(
      <MemoryRouter>
        <PopularGenres />
      </MemoryRouter>
    );

    const categories = [
      'Drama',
      'Historical',
      'Geography',
      'Fantasy/Fiction',
      'Action/Adventure',
      'Literature',
    ];

    categories.forEach((category) => {
      expect(screen.getByText(category)).toBeInTheDocument();
    });
  });

  it('has the correct links for each category', () => {
    render(
      <MemoryRouter>
        <PopularGenres />
      </MemoryRouter>
    );

    const categories = [
      { name: 'Drama', link: '/books' },
      { name: 'Historical', link: '/books' },
      { name: 'Geography', link: '/books' },
      { name: 'Fantasy/Fiction', link: '/books' },
      { name: 'Action/Adventure', link: '/books' },
      { name: 'Literature', link: '/books' },
    ];

    categories.forEach((category) => {
      const linkElement = screen.getByText(category.name).closest('a');
      expect(linkElement).toHaveAttribute('href', category.link);
    });
  });
});
