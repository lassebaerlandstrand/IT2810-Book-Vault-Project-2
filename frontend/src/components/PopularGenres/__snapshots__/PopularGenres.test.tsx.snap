// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopularGenres component > renders correctly 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-cursor-type: pointer;--mantine-font-family: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-dark-blue-0: #155896;--mantine-color-dark-blue-1: #124b81;--mantine-color-dark-blue-2: #0f3f6c;--mantine-color-dark-blue-3: #0c3256;--mantine-color-dark-blue-4: #092641;--mantine-color-dark-blue-5: #06192b;--mantine-color-dark-blue-6: #051321;--mantine-color-dark-blue-7: #030d16;--mantine-color-dark-blue-8: #030a11;--mantine-color-dark-blue-9: #02070b;--mantine-color-beige-0: #fcfaf7;--mantine-color-beige-1: #f9f5ef;--mantine-color-beige-2: #f6f0e7;--mantine-color-beige-3: #f2eade;--mantine-color-beige-4: #eee4d3;--mantine-color-beige-5: #e9ddc8;--mantine-color-beige-6: #e2d1b5;--mantine-color-beige-7: #dac5a1;--mantine-color-beige-8: #d4bd94;--mantine-color-beige-9: #6a5f4a;}:root[data-mantine-color-scheme="dark"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-8);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-9);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.15);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.2);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-3);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-4);--mantine-color-dark-blue-outline-hover: rgba(9, 38, 65, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-4);--mantine-color-beige-filled: var(--mantine-color-beige-8);--mantine-color-beige-filled-hover: var(--mantine-color-beige-9);--mantine-color-beige-light: rgba(226, 209, 181, 0.15);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.2);--mantine-color-beige-light-color: var(--mantine-color-beige-3);--mantine-color-beige-outline: var(--mantine-color-beige-4);--mantine-color-beige-outline-hover: rgba(238, 228, 211, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-dark-blue-text: var(--mantine-color-dark-blue-filled);--mantine-color-dark-blue-filled: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-filled-hover: var(--mantine-color-dark-blue-7);--mantine-color-dark-blue-light: rgba(5, 19, 33, 0.1);--mantine-color-dark-blue-light-hover: rgba(5, 19, 33, 0.12);--mantine-color-dark-blue-light-color: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline: var(--mantine-color-dark-blue-6);--mantine-color-dark-blue-outline-hover: rgba(5, 19, 33, 0.05);--mantine-color-beige-text: var(--mantine-color-beige-filled);--mantine-color-beige-filled: var(--mantine-color-beige-6);--mantine-color-beige-filled-hover: var(--mantine-color-beige-7);--mantine-color-beige-light: rgba(226, 209, 181, 0.1);--mantine-color-beige-light-hover: rgba(226, 209, 181, 0.12);--mantine-color-beige-light-color: var(--mantine-color-beige-6);--mantine-color-beige-outline: var(--mantine-color-beige-6);--mantine-color-beige-outline-hover: rgba(226, 209, 181, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class=""
    style="margin-block: var(--mantine-spacing-xl);"
  >
    <h2
      class="m_8a5d1357 mantine-Title-root"
      data-order="2"
      data-size="h3"
      style="--title-fw: var(--mantine-h3-font-weight); --title-lh: var(--mantine-h3-line-height); --title-fz: var(--mantine-h3-font-size); font-weight: 600;"
    >
      Popular Genres
    </h2>
    <div
      class="m_d57069b5 mantine-ScrollArea-root"
      style="--scrollarea-scrollbar-size: calc(0.375rem * var(--mantine-scale)); --sa-corner-width: 0px; --sa-corner-height: 0px; width: 100%;"
    >
      <div
        class="m_c0783ff9 mantine-ScrollArea-viewport"
        data-offset-scrollbars="x"
        data-scrollbars="xy"
        style="overflow-x: scroll; overflow-y: scroll;"
      >
        <div
          style="min-width: 100%; display: table;"
        >
          <div
            class="m_8bffd616 mantine-Flex-root __m__-r4"
            style="gap: var(--mantine-spacing-md); justify-content: space-between; padding-block: var(--mantine-spacing-xs);"
          >
            <a
              aria-label="Go to Drama"
              class="_link_103628"
              href="/books?genres=Drama"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Drama"
                  class="tabler-icon tabler-icon-masks-theater "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13.192 9h6.616a2 2 0 0 1 1.992 2.183l-.567 6.182a4 4 0 0 1 -3.983 3.635h-1.5a4 4 0 0 1 -3.983 -3.635l-.567 -6.182a2 2 0 0 1 1.992 -2.183z"
                  />
                  <path
                    d="M15 13h.01"
                  />
                  <path
                    d="M18 13h.01"
                  />
                  <path
                    d="M15 16.5c1 .667 2 .667 3 0"
                  />
                  <path
                    d="M8.632 15.982a4.037 4.037 0 0 1 -.382 .018h-1.5a4 4 0 0 1 -3.983 -3.635l-.567 -6.182a2 2 0 0 1 1.992 -2.183h6.616a2 2 0 0 1 2 2"
                  />
                  <path
                    d="M6 8h.01"
                  />
                  <path
                    d="M9 8h.01"
                  />
                  <path
                    d="M6 12c.764 -.51 1.528 -.63 2.291 -.36"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Drama
                </p>
              </div>
            </a>
            <a
              aria-label="Go to Historical"
              class="_link_103628"
              href="/books?genres=Historical"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Historical"
                  class="tabler-icon tabler-icon-hourglass-empty "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6 20v-2a6 6 0 1 1 12 0v2a1 1 0 0 1 -1 1h-10a1 1 0 0 1 -1 -1z"
                  />
                  <path
                    d="M6 4v2a6 6 0 1 0 12 0v-2a1 1 0 0 0 -1 -1h-10a1 1 0 0 0 -1 1z"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Historical
                </p>
              </div>
            </a>
            <a
              aria-label="Go to Geography"
              class="_link_103628"
              href="/books?genres=Geography"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Geography"
                  class="tabler-icon tabler-icon-globe "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 9a4 4 0 1 0 8 0a4 4 0 0 0 -8 0"
                  />
                  <path
                    d="M5.75 15a8.015 8.015 0 1 0 9.25 -13"
                  />
                  <path
                    d="M11 17v4"
                  />
                  <path
                    d="M7 21h8"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Geography
                </p>
              </div>
            </a>
            <a
              aria-label="Go to Fantasy/Fiction"
              class="_link_103628"
              href="/books?genres=Fantasy/Fiction"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Fantasy/Fiction"
                  class="tabler-icon tabler-icon-crystal-ball "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.73 17.018a8 8 0 1 1 10.54 0"
                  />
                  <path
                    d="M5 19a2 2 0 0 0 2 2h10a2 2 0 1 0 0 -4h-10a2 2 0 0 0 -2 2z"
                  />
                  <path
                    d="M11 7a3 3 0 0 0 -3 3"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Fantasy/Fiction
                </p>
              </div>
            </a>
            <a
              aria-label="Go to Adventure"
              class="_link_103628"
              href="/books?genres=Adventure"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Adventure"
                  class="tabler-icon tabler-icon-compass "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 16l2 -6l6 -2l-2 6l-6 2"
                  />
                  <path
                    d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
                  />
                  <path
                    d="M12 3l0 2"
                  />
                  <path
                    d="M12 19l0 2"
                  />
                  <path
                    d="M3 12l2 0"
                  />
                  <path
                    d="M19 12l2 0"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Adventure
                </p>
              </div>
            </a>
            <a
              aria-label="Go to Literature"
              class="_link_103628"
              href="/books?genres=Literature"
            >
              <div
                class="_categoryCard_103628 m_1b7284a3 mantine-Paper-root"
              >
                <svg
                  aria-label="Literature"
                  class="tabler-icon tabler-icon-pencil "
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
                  />
                  <path
                    d="M13.5 6.5l4 4"
                  />
                </svg>
                <p
                  class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                >
                  Literature
                </p>
              </div>
            </a>
          </div>
        </div>
      </div>
      <div
        class="m_c44ba933 mantine-ScrollArea-scrollbar"
        data-mantine-scrollbar="true"
        data-orientation="horizontal"
        data-state="hidden"
        style="position: absolute; --sa-thumb-width: 18px;"
      />
      <div
        class="m_c44ba933 mantine-ScrollArea-scrollbar"
        data-mantine-scrollbar="true"
        data-orientation="vertical"
        data-state="hidden"
        style="position: absolute; --sa-thumb-height: 18px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
